### ISSUE TITLE
Amazon Echo skill for AudiobookShelf

### ISSUE THREAD

**Issue by lukc82** [advplyr/audiobookshelf#316]
Hi,

It would be great to be able to control the audiobookshelf using Alexa :)

Thanks,
Luke

**Comment by advplyr** [comment]
This sounds like a fun project. I just skimmed through the documentation since I'm not familiar with Alexa.

Have you used Alexa for other self-hosted media servers? If so, I'm curious how the setup is handled. How does Alexa get your server address and login credentials for the media server?

**Comment by lukc82** [comment]
Yes. I'm using MyMedia for Alexa (https://www.mymediaalexa.com/) and I'm
surprised that since the launch of this project, it has been the only
mature alternative for listeners of self hosted audiobook files. Especially
for kids it's a wonderful tool. One word or  sentence and there you go -
your audiobook continues :)

**Comment by mx03** [comment]
Alexa needs a public audiobookshelf instance, additional it needs a central system ( e.g. audiobookshelf.org with oauth thats linked to the instance of the user).
An alternative is that everyone deploy an skill with the alexa developer console that direct linked to the audiobookshelf instance.

But alexa always need the audiofiles on an public url even the abs and the alexa is on the intranet.

**Comment by advplyr** [comment]
Yeah I thought that might be the case. Although if Alexa knew your server url and audiobookshelf login then I don't see why it would require a central server.

**Comment by mx03** [comment]
Yes thats the problem, alexa don't know the server url, if you publish an alexa skill you can link your amazon account with the server url that's in the skill.

Problem you can't pass the url to the server as a alexa user. There are two options:
- URL is bound to the skill that's in the amazon skill store (you activate the skill, and link the account with the central system thats linked to the custom user abs url, example: https://www.mymediaalexa.com )
- Sourcecode the alexa skill, user can create an amazon dev account and add the skill (hard to use, example: [Kanzi: Alexa Integration With Kodi](https://github.com/m0ngr31/kanzi))

**Comment by advplyr** [comment]
Sounds like this one will be something to look at again once the project is more mature.

I'm open to hosting a central server if it allows us to add certain features as long as it is opt-in only and transparent.

**Comment by oboe45** [comment]
Very pleased with Audiobookshelf - works perfectly for me except for the lack of Alexa or Siri operability.  Happy to contribute as a tester, but have no current skills to offer, unless you can make use of an IBM 360 assembler programmer from the 60’s!

**Comment by adocampo** [comment]
Just my two cents, I'm not a developer or whatsoever, but I set up a Lambda function to manage my Home Assistant, so I think there is no need at all for a "central server", if we, the users, have the steps to configure our own AWS Developer account and create a program with the free tier with a lambda function which connects to our ABS instance. 
We will need, besides the logic of the lambda function, which can be done in python or another programming language, a domain (I guess it can be done with free domains like no-ip.org) and publish ABS via https (which can be get with letsencrypt).
It would be great to let alexa play any book or podcast on our ABS, and start reading where it left (MyMediaForAlexa doesn't do that).

**Comment by zimler** [comment]
This would be a really cool addition to an already amazing project.

**Comment by yann-papouin** [comment]
I've already made a self-hosted python's Alexa skill to play my children bed-stories and since I've now started self-hosting audiobookshelf (amazing app by the way) I was looking for a way to replace my http-files backend with it.

I took a look on https://api.audiobookshelf.org but I didn't found a way to get all titles/authors in the library (`api/items/batch/get` ?) or to make a content fuzzy search: convert user query to phonetic and compare it with all titles/authors (stored in phonetic too) using a [levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance).

**Comment by dathbe** [comment]
I too would love this functionality.  Plex is able to play self-hosted files through Alexa, though they obviously have a central point they can direct Alexa to and then distribute out to the specific server based on login information.

Some creative folks have come up with a way you can develop your own personal Alexa skill to link with the Subsonic API.  https://github.com/rosskouk/asknavidrome .  This could presumably be done for ABS as well.  

The challenge would be, as others have pointed out, scaling this to a single production app that works for everyone's self-hosted server.  

I don't have a lot of experience developing Alexa skills (I was able to get the asknavidrome skill working, but that's it).  I know you can log in to skills.  So it would presumably be possible to keep a central database of user logins that associate with self-hosted urls that allow connection to a skill.  This has some privacy (and hosting) implications.  So it would be interesting to know if, instead of just logging in, it is possible to develop a production alexa skill that requires configuration with a server address...like we do with the iOS app now.  Not sure.

**Comment by DavidPesticcio** [comment]
Perhaps Alexa could be decoupled from the ABS server by using some no-code like Zapier in between, additionally using the cloudflared tool to expose the ABS server for added security/proivacy...?

**Comment by oboe45** [comment]
I would love to see a skill that works for ABS access - I asked about this
several times and the feedback was that it was relatively simple, but no
one actually provided the "simple" solution.  Good luck, happy to test any
solution provided.

**Comment by dathbe** [comment]
It's not the easiest thing in the world to do because there's no central hub to register with Amazon (e.g., audible.com).  It is possible, but it generally requires getting an Amazon dev account and setting up your own integration.  That's cumbersome and has its own limitations.  There are some theoretical workarounds, but it's not as straightforward as you might assume.  Some folks have tried it with music and Navidrome.  It works, but not really well.

**Comment by amwmedia** [comment]
The problem is made even more difficult when you consider that the skill would need to be setup for every account that wants to listen on Alexa using ABS. Meaning that it can't just be setup once and used by my nieces and nephews... we would need to setup a separate skill for THEM as well.

What I'm planning to build (if I can get the darn thing to play on devices) is a system that would make it simple to install the skill and link it to your ABS instance and credentials. This wouldn't be an OSS project but rather a service for a minimal cost. Hopefully something like this would be welcomed by the ABS community even though it goes against the "self-hosted", free, and open source nature of the project.

**Comment by oboe45** [comment]
Depends on the price, but likely to subscribe here

**Comment by P-Hoep** [comment]
Hm, actually have something like this close to ready for beta testing (assuming amazon lets me get that far). Nothing super fancy but seems to work alright. Having difficulty with fast forwarding/skipping chapters though (probably need to rewrite some of the url handling or find another stream option) Could probably clean up and share code if anyone is interested. It's fairly basic rn. Am using 0auth to store credentials and have it linked to the app.

**Comment by amwmedia** [comment]
@P-Hoep I'd love to take a look. I'd be very interested to see how you got the stream to play on an echo device. I'm sort of stuck there right now. It plays fine on the Alexa app but I get nothing on a real device. 😭

**Comment by oboe45** [comment]
Cool, I'll check it out!

**Comment by amwmedia** [comment]
@mx03 I'd be happy to take a look. Interesting info about the file size. I might do some testing to verify if that's actually the issue with the playback issues

**Comment by mxdalchau** [comment]
nice ideas .. is there a solution for the problem?

**Comment by amwmedia** [comment]
There are some open source projects out there (some referenced here) that will work. I'm working on a project to solve this. I'll post here once I'm ready for beta testers 😉

**Comment by adocampo** [comment]
> There are some open source projects out there (some referenced here) that will work. I'm working on a project to solve this. I'll post here once I'm ready for beta testers 😉

I would be glad to betatest it ;)

**Comment by alaenghits** [comment]
> There are some open source projects out there (some referenced here) that will work. I'm working on a project to solve this. I'll post here once I'm ready for beta testers 😉

Would love to see this.

**Comment by sdholden28** [comment]
> There are some open source projects out there (some referenced here) that will work. I'm working on a project to solve this. I'll post here once I'm ready for beta testers 😉

I'll add to the chorus. Would love to see this.

**Comment by eastoncrafter** [comment]
Another!
Willing to beta test, have quite a few echo's and the dev console

**Comment by ajlennon** [comment]
Yes please !!!

**Comment by Ngbcoder** [comment]
I can beta test too. i have multiple alexa and one at a diffrent home.

**Comment by neuther** [comment]
I have plexamp and it isn't good for paying audiobooks. I love audiobookshelf but I don't like that I don't have my books when I'm not at home. I was wondering if there's a way to have audiobookshelf use my plexamp library?

**Comment by Blackspell01** [comment]
> There are some open source projects out there (some referenced here) that will work. I'm working on a project to solve this. I'll post here once I'm ready for beta testers 😉

what happened?

**Comment by Blackspell01** [comment]
@mx03 your skill starts any audiobook from the beginning so its useless for me unfortunately...

**Comment by amwmedia** [comment]
> > There are some open source projects out there (some referenced here) that will work. I'm working on a project to solve this. I'll post here once I'm ready for beta testers 😉
> 
> what happened?

I got busy unfortunately. Hoping to resume work on this sometime this year. I'll report back here when I have something.

**Comment by mx03** [comment]
There where some issues that blocked me and as I didnt have time i didnt looked into it.

Biggest problem was/is that if you created a hls session it regenerated the complete hls stream and that breaks something sometimes in alexa, depending how long it takes to generate the stream (results: starting from the beginning or not playing the stream at all)

Best solution would be another hls stream caching and it only get deleted after some time or if the max cache size is reached.

**Comment by Blackspell01** [comment]
Why isn't it possible to build a proxy that only serves the hls parts from the current chapter on respectively?

**Comment by mx03** [comment]
> Why isn't it possible to build a proxy that only serves the hls parts from the current chapter on respectively?

If i remember correct abs already start encoding in the range of the last progress first, hls has e.g. 1000 output-[NUMBER].ts files and if you already at 50% it start at output-4xx.ts first, but alexa had a issue with it because it always read some output-[0-10].ts regardless at which progess you are, but the files arent ready (or maybe i did something wrong)

Main problem in my eyes is, if i start a session for a book with forcetranscode (hls) if there is already a existing hls stream for this entry it deletes it and generate again. (Its a useless additional transcode)

**Comment by Blackspell01** [comment]
hmm, okay, so recreating a m3u8 file that starts at 50% and also renames the ts parts to 0 wouldnt help either in this case?

**Comment by mx03** [comment]
@Blackspell01 Can you explain how you tried/setup my skill with everything around?

**Comment by Blackspell01** [comment]
skill hosted on Alexa and your proxy in a lil docker container 

FROM node:20-slim
WORKDIR /app
COPY package.json ./
RUN npm install --omit=dev
COPY index.js  ./
CMD ["node", "index.js"]
